File: app/src/main/java/com/gago/david/myland/Adapters/ItemTypeAdapter.java
package com.gago.david.myland.Adapters;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.gago.david.myland.Models.PlantTypeObject;
import com.gago.david.myland.R;
import com.gago.david.myland.SettingsFragment.OnListFragmentInteractionListener;

import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

/**
 * {@link RecyclerView.Adapter} that can display a {@link PlantTypeObject} and makes a call to the
 * specified {@link OnListFragmentInteractionListener}.
 */
public class ItemTypeAdapter extends RecyclerView.Adapter<ItemTypeAdapter.ViewHolder> {

    private final List<PlantTypeObject> mValues;
    private final OnListFragmentInteractionListener mListener;

    public ItemTypeAdapter(List<PlantTypeObject> items, OnListFragmentInteractionListener listener) {
        mValues = items;
        mListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.plant_type_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mValues.get(position);
        holder.name.setText(mValues.get(position).name);
        holder.icon.setImageResource(mValues.get(position).icon);
        holder.icon.setColorFilter(Color.parseColor(mValues.get(position).color));
        holder.remove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AlertDialog.Builder alertDialog = new AlertDialog.Builder(view.getRootView().getContext());
                alertDialog.setTitle(holder.mItem.name);
                alertDialog.setMessage(R.string.remove_item);

                ColorFilter filter = new PorterDuffColorFilter(Color.parseColor(holder.mItem.color), PorterDuff.Mode.SRC_IN);
                Drawable icon = view.getRootView().getContext().getResources().getDrawable(holder.mItem.icon);
                icon.setColorFilter(filter);
                alertDialog.setIcon(icon);

                alertDialog.setPositiveButton(R.string.yes,
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                mListener.removeItem(holder.mItem);
                                mValues.remove(holder.mItem);
                                notifyDataSetChanged();
                            }
                        });

                alertDialog.setNegativeButton(R.string.no,
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.cancel();
                            }
                        });
                alertDialog.show();
            }
        });

        holder.mView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (null != mListener) {
                    // Notify the active callbacks interface (the activity, if the
                    // fragment is attached to one) that an item has been selected.
                    mListener.selectItem(holder.mItem);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView name;
        public final CircleImageView icon;
        public final CircleImageView remove;
        public PlantTypeObject mItem;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            name = view.findViewById(R.id.plant);
            icon = view.findViewById(R.id.icon);
            remove = view.findViewById(R.id.remove);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + name.getText() + "'";
        }
    }
}

File: app/src/main/java/com/gago/david/myland/Adapters/ItemTypeAdapter.kt
package com.gago.david.myland.adapters

import android.app.AlertDialog
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.gago.david.myland.R
import com.gago.david.myland.SettingsFragment
import com.gago.david.myland.models.PlantTypeObject
import de.hdodenhof.circleimageview.CircleImageView

/**
 * [RecyclerView.Adapter] that can display a [PlantTypeObject] and makes a call to the
 * specified [OnListFragmentInteractionListener].
 */
class ItemTypeAdapter(private val mValues: MutableList<PlantTypeObject>, private val mListener: SettingsFragment.OnListFragmentInteractionListener?) : RecyclerView.Adapter<ItemTypeAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.plant_type_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mItem = mValues[position]
        holder.name.text = mValues[position].name
        holder.icon.setImageResource(mValues[position].icon)
        holder.icon.setColorFilter(Color.parseColor(mValues[position].color))
        holder.remove.setOnClickListener { view ->
            val alertDialog = AlertDialog.Builder(view.rootView.context)
            alertDialog.setTitle(holder.mItem!!.name)
            alertDialog.setMessage(R.string.remove_item)
            val filter: ColorFilter = PorterDuffColorFilter(Color.parseColor(holder.mItem!!.color), PorterDuff.Mode.SRC_IN)
            val icon = view.rootView.context.resources.getDrawable(holder.mItem!!.icon)
            icon.colorFilter = filter
            alertDialog.setIcon(icon)
            alertDialog.setPositiveButton(R.string.yes
            ) { dialog, which ->
                mListener!!.removeItem(holder.mItem)
                mValues.remove(holder.mItem!!)
                notifyDataSetChanged()
            }
            alertDialog.setNegativeButton(R.string.no
            ) { dialog, which -> dialog.cancel() }
            alertDialog.show()
        }
        holder.mView.setOnClickListener { mListener?.selectItem(holder.mItem) }
    }

    override fun getItemCount(): Int {
        return mValues.size
    }

    inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val name: TextView
        val icon: CircleImageView
        val remove: CircleImageView
        var mItem: PlantTypeObject? = null
        override fun toString(): String {
            return super.toString() + " '" + name.text + "'"
        }

        init {
            name = mView.findViewById(R.id.plant)
            icon = mView.findViewById(R.id.icon)
            remove = mView.findViewById(R.id.remove)
        }
    }
}